File: problems/famous_algorithms/dijkstras_algorithm/python/dijkstras_algorithm.py
import heapq


def init_distance_and_queue(
    start: int, n: int
) -> tuple[list[float], list[tuple[float, int]]]:
    distances = [float("inf")] * n
    distances[start] = 0
    pq = [(0, start)]
    return distances, pq


def relax_edges(
    current_vertex: int,
    current_distance: float,
    edges: list[list[tuple[int, int]]],
    distances: list[float],
    pq: list[tuple[float, int]],
    previous: list[int],
) -> None:
    for edge in edges[current_vertex]:
        if len(edge) != 2:
            continue
        neighbor, weight = edge
        new_distance = current_distance + weight

        if new_distance < distances[neighbor]:
            distances[neighbor] = new_distance
            heapq.heappush(pq, (new_distance, neighbor))
            previous[neighbor] = current_vertex


def dijkstra(
    start: int, edges: list[list[tuple[int, int]]]
) -> tuple[list[float], list[int]]:
    distances, pq = init_distance_and_queue(start, len(edges))
    previous = [-1] * len(edges)

    while pq:
        current_distance, current_vertex = heapq.heappop(pq)

        if current_distance > distances[current_vertex]:
            continue

        relax_edges(current_vertex, current_distance, edges, distances, pq, previous)

    return distances, previous


def convert_unreachable_to_negative_one(distances: list[float]) -> list[int]:
    return [-1 if x == float("inf") else x for x in distances]


def dijkstra_algorithm(
    start: int, edges: list[list[tuple[int, int]]]
) -> tuple[list[int], list[int]]:
    distances, previous = dijkstra(start, edges)
    distances = convert_unreachable_to_negative_one(distances)
    return distances, previous

File: problems/famous_algorithms/dijkstras_algorithm/js/dijkstras_algorithm.js
import { MinPriorityQueue } from "@datastructures-js/priority-queue";

export const dijkstrasAlgorithm = (() => {
  const initDistanceAndQueue = (start, n, distances, pq) => {
    distances.length = n;
    distances.fill(Infinity);
    distances[start] = 0;
    pq.enqueue({ element: start, priority: 0 });
  };

  const relaxEdges = (currentVertex, currentDistance, edges, pq, distances) => {
    if (!Array.isArray(edges[currentVertex])) return;

    for (const edge of edges[currentVertex]) {
      const [vertex, weight] = edge;

      const newDistance = currentDistance + weight;

      if (newDistance < distances[vertex]) {
        distances[vertex] = newDistance;
        pq.enqueue({ element: vertex, priority: newDistance });
      }
    }
  };

  const dijkstra = (start, edges, distances) => {
    const n = edges.length;
    const pq = new MinPriorityQueue();

    initDistanceAndQueue(start, n, distances, pq);

    while (!pq.isEmpty()) {
      const { element: currentVertex, priority: currentDistance } =
        pq.dequeue();
      if (currentDistance > distances[currentVertex]) continue;

      relaxEdges(currentVertex, currentDistance, edges, pq, distances);
    }
  };

  const convertUnreachableToNegativeOne = (distances) => {
    for (let i = 0; i < distances.length; i++) {
      if (distances[i] === Infinity) distances[i] = -1;
    }
  };

  return (start, edges) => {
    const distances = [];
    dijkstra(start, edges, distances);
    convertUnreachableToNegativeOne(distances);
    return distances;
  };
})();
